gio/src/task.rs

```diff
@@ -99,44 +304,23 @@ impl Task {
             let task_func: Box_<Q> = Box::from_raw(user_data as *mut _);
             task_func(
                 task.as_ref(),
-                source_object.as_ref().as_ref(),
+                source_object.as_ref().as_ref().map(|s| s.unsafe_cast_ref()),
                 cancellable.as_ref().as_ref(),
             );
         }
 
-        let task_func = trampoline::<Q>;
+        let task_func = trampoline::<V, S, Q>;
         unsafe {
             ffi::g_task_run_in_thread(self.to_glib_none().0, Some(task_func));
         }
     }
+}
 
-    pub fn return_value(&self, result: &glib::Value) {
-        unsafe extern "C" fn value_free(value: *mut c_void) {
-            let _: glib::Value = from_glib_full(value as *mut glib::gobject_ffi::GValue);
-        }
+unsafe impl<V: ValueType + Send> Send for Task<V> {}
```

Comment:
Same as above, this also would have to be `Sync` then (is it?). If it is not then also `run_in_task()` needs some changes as that effectively uses the task from a different thread.

AFAIU calling `return_value()` and `propagate_value()` from two threads at once is not thread-safe, and that would be a problem. You will probably need to add some more types here to allow calling these two functions only from the other thread and not from elsewhere.

----

Another problem here is that it would only be `Send` if the store callback (passed to `new()`) is also `Send`.

I think you'll need to duplicate some functions (`new()`, `connect_*()`) into a variant with `ValueType` and `ValueType + Send` so the trait bounds are fitting.